For a partitial function defined inside a value class 1 + 2 == 2.

```
class Foo(val a: Int) extends AnyVal {
  def foo = { {case x => x + a}: PartialFunction[Int, Int]}
}

object Test extends dotty.runtime.LegacyApp {
  val x = new Foo(1).foo.apply(2)
  assert(x == 3, x)
}
```
```
[info] % dotc t7571.scala
[info] % java \
[info]   -classpath \
[info]   /Users/dark/workspace/dotty/tests/partest-generated/run/t7571-run.obj:/Users/dark/workspace/dotty/target/scala-2.11/dotty_2.11-0.1-SNAPSHOT.jar:/Users/dark/.ivy2/cache/me.d-d/scala-compiler/jars/scala-compiler-2.11.5-20150506-175515-8fc7635b56.jar:/Users/dark/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar:/Users/dark/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar \
[info]   -Dfile.encoding=UTF-8 \
[info]   -Djava.library.path=/Users/dark/workspace/dotty/tests/partest-generated/run \
[info]   -Dpartest.output=/Users/dark/workspace/dotty/tests/partest-generated/run/t7571-run.obj \
[info]   -Dpartest.lib=/Users/dark/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar \
[info]   -Dpartest.reflect=/Users/dark/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar \
[info]   -Dpartest.comp=/Users/dark/workspace/dotty/target/scala-2.11/dotty_2.11-0.1-SNAPSHOT.jar \
[info]   -Dpartest.cwd=/Users/dark/workspace/dotty/tests/partest-generated/run \
[info]   -Dpartest.test-path=/Users/dark/workspace/dotty/tests/partest-generated/run/t7571.scala \
[info]   -Dpartest.testname=t7571 \
[info]   -Djavacmd=java \
[info]   -Djavaccmd=javac \
[info]   -Duser.language=en \
[info]   -Duser.country=US \
[info]   scala.tools.nsc.MainGenericRunner \
[info]   -usejavacp \
[info]   Test \
[info]   jvm > t7571-run.log
[info] java.lang.AssertionError: assertion failed: 2
[info] 	at scala.Predef$.assert(Predef.scala:165)
[info] 	at Test$.<init>(t7571.scala:14)
[info] 	at Test$.<clinit>(t7571.scala)
[info] 	at Test.main(t7571.scala)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info] 	at java.lang.reflect.Method.invoke(Method.java:497)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$$anonfun$run$1.apply(ScalaClassLoader.scala:70)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:101)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:70)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:101)
[info] 	at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:22)
[info] 	at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:39)
[info] 	at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:29)
[info] 	at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:39)
[info] 	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:65)
[info] 	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
[info] 	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
[info] 	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
[info] 	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

--- a/src/dotty/tools/dotc/ast/tpd.scala
+++ b/src/dotty/tools/dotc/ast/tpd.scala
@@ -76,8 +76,13 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def Block(stats: List[Tree], expr: Tree)(implicit ctx: Context): Block =
     ta.assignType(untpd.Block(stats, expr), stats, expr)
 
-  def maybeBlock(stats: List[Tree], expr: Tree)(implicit ctx: Context): Tree =
-    if (stats.isEmpty) expr else Block(stats, expr)
+  /** Join `stats` in front of `expr` creating a new block if necessary */
+  def seq(stats: List[Tree], expr: Tree)(implicit ctx: Context): Tree =
+    if (stats.isEmpty) expr
+    else expr match {
+      case Block(estats, eexpr) => cpy.Block(expr)(stats ::: estats, eexpr)
+      case _ => Block(stats, expr)
+    }
 
   def If(cond: Tree, thenp: Tree, elsep: Tree)(implicit ctx: Context): If =
     ta.assignType(untpd.If(cond, thenp, elsep), thenp, elsep)

--- a/src/dotty/tools/dotc/transform/ExplicitOuter.scala
+++ b/src/dotty/tools/dotc/transform/ExplicitOuter.scala
@@ -136,7 +136,9 @@ object ExplicitOuter {
 
   /** A new outer accessor or param accessor */
   private def newOuterSym(owner: ClassSymbol, cls: ClassSymbol, name: TermName, flags: FlagSet)(implicit ctx: Context) = {
-    ctx.newSymbol(owner, name, Synthetic | flags, cls.owner.enclosingClass.typeRef, coord = cls.coord)
+    val target = cls.owner.enclosingClass.typeRef
+    val info = if (flags.is(Method)) ExprType(target) else target
+    ctx.newSymbol(owner, name, Synthetic | flags, info, coord = cls.coord)
   }
 
   /** A new param accessor for the outer field in class `cls` */
@@ -302,7 +304,7 @@ object ExplicitOuter {
         val outerAccessorCtx = ctx.withPhaseNoLater(ctx.lambdaLiftPhase) // lambdalift mangles local class names, which means we cannot reliably find outer acessors anymore
         ctx.log(i"outer to $toCls of $tree: ${tree.tpe}, looking for ${outerAccName(treeCls.asClass)(outerAccessorCtx)} in $treeCls")
         if (treeCls == toCls) tree
-        else loop(tree select outerAccessor(treeCls.asClass)(outerAccessorCtx))
+        else loop(tree.select(outerAccessor(treeCls.asClass)(outerAccessorCtx)).ensureApplied)
       }
       ctx.log(i"computing outerpath to $toCls from ${ctx.outersIterator.map(_.owner).toList}")
       loop(This(ctx.owner.enclosingClass.asClass))

--- a/src/dotty/tools/dotc/transform/LambdaLift.scala
+++ b/src/dotty/tools/dotc/transform/LambdaLift.scala
@@ -384,14 +384,23 @@ class LambdaLift extends MiniPhase with IdentityDenotTransformer { thisTransform
     private def addFreeParams(tree: Tree, proxies: List[Symbol])(implicit ctx: Context, info: TransformerInfo): Tree = proxies match {
       case Nil => tree
       case proxies =>
+        val sym = tree.symbol
         val ownProxies =
-          if (!tree.symbol.isConstructor) proxies
-          else proxies.map(_.copy(owner = tree.symbol, flags = Synthetic | Param))
+          if (!sym.isConstructor) proxies
+          else proxies.map(_.copy(owner = sym, flags = Synthetic | Param))
         val freeParamDefs = ownProxies.map(proxy =>
           transformFollowingDeep(ValDef(proxy.asTerm).withPos(tree.pos)).asInstanceOf[ValDef])
+        def proxyInit(field: Symbol, param: Symbol) =
+          transformFollowingDeep(ref(field).becomes(ref(param)))
+        def copyParams(rhs: Tree) = {
+          ctx.log(i"copy params ${proxies.map(_.showLocated)}%, %, own = ${ownProxies.map(_.showLocated)}%, %")
+          seq((proxies, ownProxies).zipped.map(proxyInit), rhs)
+        }
         tree match {
           case tree: DefDef =>
-            cpy.DefDef(tree)(vparamss = tree.vparamss.map(freeParamDefs ++ _))
+            cpy.DefDef(tree)(
+                vparamss = tree.vparamss.map(freeParamDefs ++ _),
+                rhs = if (sym.isPrimaryConstructor) copyParams(tree.rhs) else tree.rhs)
           case tree: Template =>
             cpy.Template(tree)(body = freeParamDefs ++ tree.body)
         }

--- a/tests/run/i659.scala
+++ b/tests/run/i659.scala
@@ -1,12 +1,26 @@
 class Foo(val a: Int) {
-  //def foo = { {case x => x + a}: PartialFunction[Int, Int]}
+  def foo = { {case x => x + a}: PartialFunction[Int, Int]}
   class Bar { def result(x: Int) = x + a }
   def bar = new Bar
 }
 
+class VFoo(val a: Int) extends AnyVal {
+  def foo = { {case x => x + a}: PartialFunction[Int, Int]}
+}
+
 object Test extends dotty.runtime.LegacyApp {
-  //val x = new Foo(1).foo.apply(2)
-  val y = new Foo(1).bar.result(2)
-  assert(y == 3, y)
-  //assert(x == 3, x)
+
+  def Foo(a: Int) = {
+    class Bar { def result(x: Int) = x + a }
+    new Bar().result(2)
+  }
+
+  val x1 = new Foo(1).bar.result(2)
+  assert(x1 == 3, s"x1 = $x1")
+  val x2 = Foo(1)
+  assert(x2 == 3, s"x2 = $x2")
+  val x3 = new Foo(1).foo.apply(2)
+  assert(x3 == 3, s"x3 = $x3")
+  val x4 = new VFoo(1).foo.apply(2)
+  assert(x4 == 3, s"x4 = $x4")
 }

--- a/tests/run/i659.scala
+++ b/tests/run/i659.scala
@@ -1,0 +1,12 @@
+class Foo(val a: Int) {
+  //def foo = { {case x => x + a}: PartialFunction[Int, Int]}
+  class Bar { def result(x: Int) = x + a }
+  def bar = new Bar
+}
+
+object Test extends dotty.runtime.LegacyApp {
+  //val x = new Foo(1).foo.apply(2)
+  val y = new Foo(1).bar.result(2)
+  assert(y == 3, y)
+  //assert(x == 3, x)
+}

